de.metas.shipper.gateway.dhl/src/main/java/de/metas/shipper/gateway/dhl/DhlDeliveryOrderRepository.java

```diff
@@ -0,0 +1,359 @@
+					// (2.2.4.2)
+					//noinspection ConstantConditions
+					shipmentOrder.setDHL_Receiver_Email(deliveryContact != null ? deliveryContact.getEmailAddress() : null);
+					//noinspection ConstantConditions
+					shipmentOrder.setDHL_Receiver_Phone(deliveryContact != null ? deliveryContact.getPhoneAsStringOrNull() : null);
+				}
+
+				{
+					// (2.2.2) Shipper aka Pickup
+					final Address pickupAddress = deliveryOrder.getPickupAddress();
+					// (2.2.2.1)
+					shipmentOrder.setDHL_Shipper_Name1(pickupAddress.getCompanyName1());
+					shipmentOrder.setDHL_Shipper_Name2(pickupAddress.getCompanyName2());
+					// (2.2.2.2)
+					shipmentOrder.setDHL_Shipper_StreetName1(pickupAddress.getStreet1());
+					shipmentOrder.setDHL_Shipper_StreetName2(pickupAddress.getStreet2());
+					shipmentOrder.setDHL_Shipper_StreetNumber(pickupAddress.getHouseNo());
+					shipmentOrder.setDHL_Shipper_ZipCode(pickupAddress.getZipCode());
+					shipmentOrder.setDHL_Shipper_City(pickupAddress.getCity());
+					// (2.2.2.2.8)
+					shipmentOrder.setDHL_Shipper_CountryISO2Code(pickupAddress.getCountry().getAlpha2());
+					shipmentOrder.setDHL_Shipper_CountryISO3Code(pickupAddress.getCountry().getAlpha3());
+				}
+
+				InterfaceWrapperHelper.save(shipmentOrder);
+				{
+					// (2.1) The id column (I_DHL_ShipmentOrder_ID) is used as ShipmentOrder.sequenceNumber since it's unique
+					// nothing to persist here, but must be filled when retrieving the PO
+				}
+			} // fori loop
+		}
+		return shipmentOrderRequest;
+	}
+
+	private void toUpdateShipmentOrderRequestPO(@NonNull final DeliveryOrder deliveryOrder)
+	{
+		final I_DHL_ShipmentOrderRequest shipmentOrderRequest = InterfaceWrapperHelper.load(deliveryOrder.getRepoId(), I_DHL_ShipmentOrderRequest.class);
+
+		for (final DeliveryPosition deliveryPosition : deliveryOrder.getDeliveryPositions()) // only a single delivery position should exist
+		{
+			final ImmutableList<Integer> packageIdsAsList = deliveryPosition.getPackageIds().asList();
+			for (int i = 0; i < deliveryPosition.getNumberOfPackages(); i++)
+			{
+				final DhlCustomDeliveryData customDeliveryData = (DhlCustomDeliveryData)deliveryOrder.getCustomDeliveryData();
+
+				final I_DHL_ShipmentOrder shipmentOrder = getShipmentOrderByRequestIdAndPackageId(deliveryOrder.getRepoId(), packageIdsAsList.get(i));
+				//noinspection ConstantConditions
+				shipmentOrder.setawb(customDeliveryData.getSequenceNumberToAWB().get(Integer.toString(shipmentOrder.getDHL_ShipmentOrder_ID())));
+				shipmentOrder.setPdfLabelData(customDeliveryData.getSequenceNumberToPdfLabel().get(Integer.toString(shipmentOrder.getDHL_ShipmentOrder_ID())));
+
+				InterfaceWrapperHelper.save(shipmentOrder);
+			}
+		}
+	}
+
+	private static I_DHL_ShipmentOrder getShipmentOrderByRequestIdAndPackageId(final int requestId, final int packageId)
+	{
+		return Services.get(IQueryBL.class)
+				.createQueryBuilderOutOfTrx(I_DHL_ShipmentOrder.class)
```

Comment:
why out of trx?

Reply:
no particular reason, i think this was the first autocompleted result

Reply:
pls pay attention to this because it's uber important.
in ur case, i think u shall retrieve in transaction.